content/blog/2018-06-07-when-to-use-derived-state.md

```diff
@@ -0,0 +1,331 @@
+  }
+
+  handleChange = event => this.setState({ email: event.target.value });
+}
+```
+
+This design makes the component a little easier to use, but it also has a downside. Let's say that our "edit profile" form could be used to edit multiple users. Whenever a new user is selected, the form should reset the email field. If our `EmailInput` is uncontrolled, it will ignore any changes to `props.defaultEmail`. So how could we support this use case?
+
+#### Option 1: Reset uncontrolled component with a `key`
+
+One approach would be to use a special React prop called `key`. React uses this property to decide whether to [_update_ a component instance or _create_ a new one](/docs/reconciliation.html#keys). In our case, we could use an attribute like a user id to recreate the email input any time a new user is selected. Each time the `EmailInput` is recreated, it will "reset" its state to the value specified by the `defaultEmail` prop.
+
+```js
+<EmailInput
+  defaultEmail={this.props.user.email}
+  key={this.props.user.id}
+/>
+```
+
+#### Option 2: Reset uncontrolled component with an id prop
+
+Recreating this simple example component is cheap, but what if the component was expensive to initialize? Or what if it had other state that we _did not want to reset_?
+
+A slight variation on the above approach would be to watch for changes in a prop like "id" and use it to reset state:
+
+```js
+class EmailInput extends Component {
+  state = {
+    email: this.props.defaultEmail,
+    prevPropsID: this.props.id
+  };
+
+  static getDerivedStateFromProps(props, state) {
+    // Any time the current user changes,
+    // Reset any parts of state that are tied to that user.
+    // In this simple example, that's just the email.
+    if (props.id !== state.prevPropsID) {
+      return {
+        prevPropsID: props.id,
+        email: props.email
+      };
+    }
+    return null;
+  }
+
+  // ...
+}
+```
+
+This approach can scale better with multiple state values, since fewer comparisons are required. It also provides the flexibility to only reset parts of our component's internal state.
+
+#### Option 3: Reset uncontrolled component with an instance method
+
+Not every component accepts a special prop like "id". (For example, what if we were writing a list component instead, and wanted to accept a scroll offset prop?) For cases like this, we could expose an instance method to imperatively reset the internal state:
+
+```js
+class EmailInput extends Component {
+  state = {
+    email: this.props.defaultEmail
+  };
+
+  resetEmailForNewUser(newEmail) {
+    this.setState({ email: newEmail });
+  }
+
+  // ...
+}
+```
+
+The parent form component could then [use a `ref` to call this method](/docs/glossary.html#refs).
+
+> Note
+>
+> Refs can be useful when interfacing with imperative third party APIs or in certain, limited use cases like the one mentioned above, but generally we recommend you use them sparingly.
+
```

Comment:
I think it would help if there was a small recap here that gave me a quick path to choosing which option/alternative to use. For example by now I've already forgotten about controlled components but that may be the best solution for my use case.

Reply:
How about something like...
> To recap, there are a couple of options for resetting an uncontrolled component's state. Which one is best depends on the component you are writing. Here is a rough guide:
1. To reset _all internal state_, use the `key` attribute.
2. To reset _some internal state_, watch for changes in a special property (e.g. `props.id`).
3. Otherwise, consider falling back to an imperative instance method called with a `ref`.

Reply:
This leaves out the (often best) option: lift the state up and deal with it explicitly in the parent. For example, a form might want to have both `draftValues` and `committedValues` as state in a parent component.

Reply:
(I'm proposing to summarize the whole "how to avoid anti-patterns" section, not just the uncontrolled one)

Reply:
Do you have proposed wording for this summary?

It can be a little hard to just insert changes like this after I've read and re-read the post so may times. I get kind of stuck in my thinking patterns.

Let me push what I have here, and then maybe you can make one last round fo suggestions.

Reply:
I will make an attempt at doing a more comprehensive summary though.

Reply:
1. **If you're trying to "mirror" a prop value in the state, you can make the component controlled, and consolidate the two diverging values in the state of some parent component.** For example, an input tracking a "draft" `state.value` that's derived from a "committed" `props.value` could instead always use `props.value` and call `props.onChange`. A form component above could then explicitly own two diverging values in its own state, such as `state.draftValue` and `state.committedValue`, pass one of them down, and reset either of them when appropriate. (Don't forget you can use a nested object if the form contains multiple fields!)

2. **If you're trying to "reset" one or more state fields when a particular unrelated prop (usually an ID) changes,** you have a few options:

(your list)

Reply:
It's not clear if you're suggesting replacing my intro paragraph (below) or inserting between it and the uncontrolled bullets.

I think my recap had a bit of context that this suggestion loses.

I'll try to combine the two.

Reply:
I meant that your list comes under the second point :-)

<img width="388" alt="screen shot 2018-06-05 at 11 02 45 pm" src="https://user-images.githubusercontent.com/810438/41005248-99ca4918-6914-11e8-91af-23bcc96cec03.png">